### Sprint 9 Project

### Summary:

### Step 1: Download and Prepare the Data. Explain the Procedure

In [1]:
#Import the necessary libraries.
import pandas as pd

from sklearn.model_selection import train_test_split

from sklearn.linear_model import LinearRegression

from sklearn.model_selection import RandomizedSearchCV

from sklearn.pipeline import Pipeline

from sklearn.preprocessing import StandardScaler

from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score

from sklearn.model_selection import cross_val_score

import numpy as np

In [2]:
#Use the pd.read_csv() method in order to create a dataframe for Region 0.
r0 = pd.read_csv('geo_data_0.csv')

In [3]:
#Use the head() method to check the first five lines of the r0 dataframe.
r0.head()

id        f0        f1        f2     product
0  txEyH  0.705745 -0.497823  1.221170  105.280062
1  2acmU  1.334711 -0.340164  4.365080   73.037750
2  409Wp  1.022732  0.151990  1.419926   85.265647
3  iJLyR -0.032172  0.139033  2.978566  168.620776
4  Xdl7t  1.988431  0.155413  4.751769  154.036647

Code Result:
It appears that the r0 dataframe has been sucessfully uploaded.

In [4]:
#Use the pd.read_csv() method in order to create a dataframe for Region 1.
r1 = pd.read_csv('geo_data_1.csv')

In [5]:
#Use the head() method to check the first five lines of the r1 dataframe.
r1.head()

id         f0         f1        f2     product
0  kBEdx -15.001348  -8.276000 -0.005876    3.179103
1  62mP7  14.272088  -3.475083  0.999183   26.953261
2  vyE1P   6.263187  -5.948386  5.001160  134.766305
3  KcrkZ -13.081196 -11.506057  4.999415  137.945408
4  AHL4O  12.702195  -8.147433  5.004363  134.766305

Code Result:
It appears that the r1 dataframe has been sucessfully uploaded.

In [6]:
#Use the pd.read_csv() method in order to create a dataframe for Region 2.
r2 = pd.read_csv('geo_data_2.csv')

In [7]:
#Use the head() method to check the first five lines of the r2 dataframe.
r2.head()

id        f0        f1        f2     product
0  fwXo0 -1.146987  0.963328 -0.828965   27.758673
1  WJtFt  0.262778  0.269839 -2.530187   56.069697
2  ovLUW  0.194587  0.289035 -5.586433   62.871910
3  q6cA6  2.236060 -0.553760  0.930038  114.572842
4  WPMUX -0.515993  1.716266  5.899011  149.600746

Code Result:
It appears that the r2 dataframe has been sucessfully uploaded.

### Data Checking

In [8]:
#Use the isna() and sum() methods to count number of missing values in the r0 dataframe.
r0.isna().sum()

id         0
f0         0
f1         0
f2         0
product    0
dtype: int64

Code Result:
There are no missing values noted in the r0 dataframe.

In [9]:
#Use the isna() and sum() methods to count number of missing values in the r1 dataframe.
r1.isna().sum()

id         0
f0         0
f1         0
f2         0
product    0
dtype: int64

Code Result:
There are no missing values noted in the r2 dataframe.

In [10]:
#Use the isna() and sum() methods to count number of missing values in the r2 dataframe.
r2.isna().sum()

id         0
f0         0
f1         0
f2         0
product    0
dtype: int64

#Use the isna() and sum() methods to count number of missing values in the r2 dataframe.

In [11]:
#Use the duplicated() and sum() methods to count the number of duplicate values in the r0 dataframe.
r0.duplicated().sum()

0

Code Result:
There are no duplicate values noted in the r0 dataframe.

In [12]:
#Use the duplicated() and sum() methods to count the number of duplicate values in the r1 dataframe.
r1.duplicated().sum()

0

Code Result:
There are no duplicate values noted in the r1 dataframe.

In [13]:
#Use the duplicated() and sum() methods to count the number of duplicate values in the r2 dataframe.
r2.duplicated().sum()

0

Code Result:
There are no duplicate values noted in the r2 dataframe.

In [14]:
#Use the info() method in order check the data types in the r0 dataframe.
r0.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


Code Result:
The data types in the r0 dataframe appear to be listed correctly.

In [15]:
#Use the info() method in order check the data types in the r1 dataframe.
r1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


Code Result:
The data types in the r1 dataframe appear to be listed correctly.

In [16]:
#Use the info() method in order check the data types in the r2 dataframe.
r2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


Code Result:
The data types in the r2 dataframe appear to be listed correctly.

### Data Splitting

In [17]:
#Prepare the features and target values with the data from the r0 dataframe.
features_r0 = r0.drop(['id', 'product'], axis=1)

target_r0 = r0['product']

In [18]:
#Use the train_test_split() method in order to split the features_r0 and target_r0 dataframes into a training and validation set 
#with a 75:25 ratio.
features_train_r0, features_valid_r0, target_train_r0, target_valid_r0 = train_test_split(features_r0, target_r0, test_size=0.25, random_state=12345)


In [19]:
#Prepare the features and target values with the data from the r1 dataframe.
features_r1 = r1.drop(['id', 'product'], axis=1)

target_r1 = r1['product']

In [20]:
#Use the train_test_split() method in order to split the features_r1 and target_r1 dataframes into a training and validation set 
#with a 75:25 ratio.
features_train_r1, features_valid_r1, target_train_r1, target_valid_r1 = train_test_split(features_r1, target_r1, test_size=0.25, random_state=12345)

In [21]:
#Prepare the features and target values with the data from the r2 dataframe.
features_r2 = r2.drop(['id', 'product'], axis=1)

target_r2 = r2['product']

In [22]:
#Use the train_test_split() method in order to split the features_r2 and target_r2 dataframes into a training and validation set 
#with a 75:25 ratio.
features_train_r2, features_valid_r2, target_train_r2, target_valid_r2 = train_test_split(features_r2, target_r2, test_size=0.25, random_state=12345)


### Step 2. Train and Test the Model for Each Region.

#### Region 0 Model

In [23]:
#Create a Linear Regression Model for Region 0 using Scaler() and LinearRegression().
r0_model = Pipeline([('scaler', StandardScaler()), ('regressor', LinearRegression())])

In [24]:
#Create a parameter grid for the LinearRegression() model.
param_grid = {'scaler__with_mean': [True, False], 'scaler__with_std': [True, False]}

In [25]:
#Use RandomizedSearchCV() in order to create a Randomized Search model in order to tune the hyperparameters for the 
#LinearRegression() model.
r0_lr_rsearch = RandomizedSearchCV(estimator=r0_model, param_distributions=param_grid, n_iter=10, cv=5, random_state=12345, n_jobs=1)

In [26]:
#Use the fit() method in order to train the RandomizedSearchCV() model with the features and target from the training set for
#Region 0.
r0_lr_rsearch.fit(features_train_r0, target_train_r0)

c:\Users\HP\anaconda3\Lib\site-packages\sklearn\model_selection\_search.py:320: UserWarning: The total space of parameters 4 is smaller than n_iter=10. Running 4 iterations. For exhaustive searches, use GridSearchCV.
  warnings.warn(


RandomizedSearchCV(cv=5,
                   estimator=Pipeline(steps=[('scaler', StandardScaler()),
                                             ('regressor',
                                              LinearRegression())]),
                   n_jobs=1,
                   param_distributions={'scaler__with_mean': [True, False],
                                        'scaler__with_std': [True, False]},
                   random_state=12345)

In [27]:
#Use the best_params_ and best_score_methods_ in order to print out the best score and the best hyperparameters for the Linear Regression model.
r0_lr_bparams = r0_lr_rsearch.best_params_

r0_lr_bscore = r0_lr_rsearch.best_score_

print("Region 0 Linear Regression Best Hyperparameters:", r0_lr_bparams)

print('Region 0 Linear Regression Best Score:', r0_lr_bscore)

Region 0 Linear Regression Best Hyperparameters: {'scaler__with_std': True, 'scaler__with_mean': True}
Region 0 Linear Regression Best Score: 0.27408653371000735


Code Result:
It appears that the best hyperparameters for the Linear Regression Model for Region 0 are:  {'scaler__with_std': True, 'scaler__with_mean': True} and the best score for the model is approximately 0.27.

In [28]:
#Use Pipeline() and LinearRegression() once again in order to recreate r0_model with the new tuned hyperparameters.
r0_model = Pipeline([('scaler', StandardScaler(with_std=True, with_mean=True,)), ('regressor', LinearRegression())])

In [29]:
#Use the cross_val_score() method in order to find the RMSE for r0_model.
r0_scores = cross_val_score(r0_model, features_train_r0, target_train_r0, cv=5, scoring='neg_mean_squared_error')

r0_rmse_scores = (-r0_scores) ** 0.5

r0_rmse = r0_rmse_scores.mean()

print("Linear Regression Model RMSE:", r0_rmse)

Linear Regression Model RMSE: 37.73238188617927


Code Result:
The RMSE for r0_model is: approximately 37.73.

In [30]:
#Use the fit() method in order to train r0_model with the features_train_r0 and target_train_r0 variables.
r0_model.fit(features_train_r0, target_train_r0)

Pipeline(steps=[('scaler', StandardScaler()),
                ('regressor', LinearRegression())])

In [31]:
#Use the predict() method on r0_model in order to make a prediction using the features_valid_r0 variable.
r0_predict = r0_model.predict(features_valid_r0)

In [32]:
#Use the mean() method on the r0_predict variable in order to calcualte the average predicted reserves for Region 0.
r0_apr = r0_predict.mean()

#Print the amount of average predicted reserves for Region 0.
print("Region 0 Average Predicted Reserves:", r0_apr)

#Print the rmse for the r0_model.
print("Linear Regression Region 0 Model RMSE", r0_rmse)

Region 0 Average Predicted Reserves: 92.59256778438035
Linear Regression Region 0 Model RMSE 37.73238188617927


Code Result:
It appears that according to the model, the average amount predicted of reserves for Region 0 is approximately 92.59 thousand barrels.
The RMSE for the Linear Regression Model for Region 0 (r0_model) is: approximately 37.73.


#### Region 1 Model

In [33]:
#Create a Linear Regression Model for Region 1 using Scaler() and LinearRegression().
r1_model = Pipeline([('scaler', StandardScaler()), ('regressor', LinearRegression())])

In [34]:
#Use RandomizedSearchCV() in order to create a Randomized Search model in order to tune the hyperparameters for the 
#LinearRegression() model.
r1_lr_rsearch = RandomizedSearchCV(estimator=r1_model, param_distributions=param_grid, n_iter=10, cv=5, random_state=12345, n_jobs=1)

In [35]:
#Use the fit() method in order to train the RandomizedSearchCV() model with the features and target from the training set for
#Region 1.
r1_lr_rsearch.fit(features_valid_r1, target_valid_r1)

c:\Users\HP\anaconda3\Lib\site-packages\sklearn\model_selection\_search.py:320: UserWarning: The total space of parameters 4 is smaller than n_iter=10. Running 4 iterations. For exhaustive searches, use GridSearchCV.
  warnings.warn(


RandomizedSearchCV(cv=5,
                   estimator=Pipeline(steps=[('scaler', StandardScaler()),
                                             ('regressor',
                                              LinearRegression())]),
                   n_jobs=1,
                   param_distributions={'scaler__with_mean': [True, False],
                                        'scaler__with_std': [True, False]},
                   random_state=12345)

In [36]:
#Use the best_params_ and best_score_methods_ in order to print out the best score and the best hyperparameters for the Linear Regression model.
r1lr_bparams = r1_lr_rsearch.best_params_
r1_lr_bscore = r1_lr_rsearch.best_score_

print("Linear Regression Region 2 Best Hyperparameters:", r1lr_bparams)

print("Linear Regression Region 2 Best Score:", r1_lr_bscore)

Linear Regression Region 2 Best Hyperparameters: {'scaler__with_std': True, 'scaler__with_mean': True}
Linear Regression Region 2 Best Score: 0.9996232414326588


Code Result:
It appears that the best hyperparameters for the Linear Regression Model for Region 1 are:  {'scaler__with_std': True, 'scaler__with_mean': True} and the best score for the model is approximately 0.99.


In [37]:
#Use Pipeline() and LinearRegression() once again in order to recreate r1_model with the new tuned hyperparameters.
r1_model = Pipeline([('scaler', StandardScaler(with_std=True, with_mean=True,)), ('regressor', LinearRegression())])

In [38]:
#Use the cross_val_score() method in order to find the RMSE for r1_model.
r1_scores = cross_val_score(r1_model, features_train_r1, target_train_r1, cv=5, scoring='neg_mean_squared_error')

r1_rmse_scores = (-r1_scores) ** 0.5

r1_rmse = r1_rmse_scores.mean()

print("Linear Regression Region 1 Model RMSE:", r1_rmse)

Linear Regression Region 1 Model RMSE: 0.8895409511809378


Code Result:
The RMSE for r1_model is: approximately 0.88.

In [39]:
#Use the fit() method in order to train r1_model with the features_train_r1 and target_train_r1 variables.
r1_model.fit(features_train_r1, target_train_r1)

Pipeline(steps=[('scaler', StandardScaler()),
                ('regressor', LinearRegression())])

In [40]:
#Use the predict() method on r1_model in order to make a prediction using the features_valid_r1 variable.
r1_predict = r1_model.predict(features_valid_r1)

In [41]:
#Use the mean() method on the r1_predict variable in order to calcualte the average predicted reserves for Region 1.
r1_apr = r1_predict.mean()

#Print the amount of average predicted reserves for Region 1.
print("Region 1 Average Predicted Reserves:", r1_apr)

#Print the rmse for the r1_model.
print("Linear Regression Region 1 Model RMSE:", r1_rmse)

Region 1 Average Predicted Reserves: 68.728546895446
Linear Regression Region 1 Model RMSE: 0.8895409511809378


Code Result:
It appears that according to the model, the average amount predicted of reserves for Region 1 is approximately 68.72 thousand barrels.
The RMSE for the Linear Regression Model for Region 1 (r1_model) is: approximately 0.88.


#### Region 2 Model

In [42]:
r2_model = Pipeline([('scaler', StandardScaler()), ('regressor', LinearRegression())])

In [43]:
r2_lr_rsearch = RandomizedSearchCV(estimator=r2_model, param_distributions=param_grid, n_iter=10, cv=5, random_state=12345, n_jobs=1)

In [44]:
r2_lr_rsearch.fit(features_train_r2, target_train_r2)

c:\Users\HP\anaconda3\Lib\site-packages\sklearn\model_selection\_search.py:320: UserWarning: The total space of parameters 4 is smaller than n_iter=10. Running 4 iterations. For exhaustive searches, use GridSearchCV.
  warnings.warn(


RandomizedSearchCV(cv=5,
                   estimator=Pipeline(steps=[('scaler', StandardScaler()),
                                             ('regressor',
                                              LinearRegression())]),
                   n_jobs=1,
                   param_distributions={'scaler__with_mean': [True, False],
                                        'scaler__with_std': [True, False]},
                   random_state=12345)

In [45]:
r2_bparams = r2_lr_rsearch.best_params_
r2_bscore = r2_lr_rsearch.best_score_

print("Linear Regression Region 2 Model Best Hyperparameters:", r2_bparams)
print("Linear Regression Model Region 2 Best Score:", r2_bscore)

Linear Regression Region 2 Model Best Hyperparameters: {'scaler__with_std': True, 'scaler__with_mean': True}
Linear Regression Model Region 2 Best Score: 0.19648225543040282


In [46]:
r2_model = Pipeline([('scaler', StandardScaler(with_std=True, with_mean=True,)), ('regressor', LinearRegression())])

In [47]:
r2_scores = cross_val_score(r2_model, features_train_r2, target_train_r2, cv=5, scoring='neg_mean_squared_error')

r2_rmse_scores = (-r2_scores) ** 0.5

r2_rmse = r2_rmse_scores.mean()

print("Linear Regression Region 3 Model RMSE:", r2_rmse)

Linear Regression Region 3 Model RMSE: 40.06576061701094


In [48]:
r2_model.fit(features_train_r2, target_train_r2)

Pipeline(steps=[('scaler', StandardScaler()),
                ('regressor', LinearRegression())])

In [49]:
r2_predict = r2_model.predict(features_valid_r2)

In [50]:
r2_apr = r2_predict.mean()

print("Region 2 Average Predicted Reserves:", r2_apr)
print("Linear Regression Model Region 2 RMSE:", r2_rmse)

Region 2 Average Predicted Reserves: 94.96504596800489
Linear Regression Model Region 2 RMSE: 40.06576061701094


3.1 Store All the Key Values in Separate Variables

In [51]:
budget = 100000000

revenue_per_barrel = 4500

cost_per_well = budget/200

3.2 Calculate the volume of reserves sufficient for developing a new well without losses

In [52]:
cost_per_well

500000.0

In [53]:
sufficient_volume = cost_per_well/revenue_per_barrel

sufficient_volume

111.11111111111111

In [54]:
r0.head()

id        f0        f1        f2     product
0  txEyH  0.705745 -0.497823  1.221170  105.280062
1  2acmU  1.334711 -0.340164  4.365080   73.037750
2  409Wp  1.022732  0.151990  1.419926   85.265647
3  iJLyR -0.032172  0.139033  2.978566  168.620776
4  Xdl7t  1.988431  0.155413  4.751769  154.036647

In [55]:
r0_valid_df = features_valid_r0.copy()
r0_valid_df['actual_valid'] = target_valid_r0
r0_valid_df['predicted'] = r0_predict

r0_valid_df.head()

f0        f1        f2  actual_valid  predicted
71751  0.948970 -0.057547  2.095727     10.038645  95.894952
80493  0.992974  0.206671 -0.142278    114.551489  77.572583
2655   1.199854 -0.563356 -1.852991    132.603635  77.892640
53233  0.691422 -0.433347  0.564974    169.072125  90.175134
91141  0.420772  0.972638  0.736190    122.325180  70.510088

In [56]:
r1_valid_df = features_valid_r1.copy()
r1_valid_df['acutal_valid'] = target_valid_r1
r1_valid_df['predicted'] = r1_predict

r1_valid_df.head()

f0         f1        f2  acutal_valid  predicted
71751  -0.371866  -1.862494  3.002210     80.859783  82.663314
80493   9.015122 -13.881455  1.995363     53.906522  54.431786
2655   -6.507568  -4.817448  1.003449     30.132364  29.748760
53233  14.560845 -10.667755  1.995175     53.906522  53.552133
91141   6.090476  -4.494723  0.013815      0.000000   1.243856

In [57]:
r2_valid_df = features_valid_r2.copy()
r2_valid_df['actual_valid'] = target_valid_r2
r2_valid_df['predicted'] = r2_predict

r2_valid_df.head()

f0        f1        f2  actual_valid   predicted
71751 -1.444717 -3.861599  2.225805     61.212375   93.599633
80493 -1.418617  1.276544 -0.976433     41.850118   75.105159
2655  -4.587649 -0.413199  1.649268     57.776581   90.066809
53233  1.871584  1.619101  4.273555    100.053761  105.162375
91141 -2.028785  4.128167  6.089547    109.897122  115.303310

#### Revenue Function

In [58]:
#Revenue Function
def calc_profit(valid_df, n_top_wells=200, barrel_price=4500, budget=100e6):
    predicted_sorted = valid_df.sort_values(by='predicted', ascending=False)
    
    top_well = predicted_sorted.head(n_top_wells)
    
    total_revenue = top_well['predicted'].sum() * barrel_price
    
    profit = total_revenue - budget
    
    return profit

In [59]:
r0_profit = calc_profit(r0_valid_df)

print(f"Region 0 Estimated Profit: ${r0_profit:.2f}") 

Region 0 Estimated Profit: $39960488.77


In [60]:
r0_reserves = r0_profit / revenue_per_barrel

print(f"Region 0 Estimated Volume of Reserves According to Predicitons: {int(r0_reserves)}")

Region 0 Estimated Volume of Reserves According to Predicitons: 8880


In [61]:
r1_profit = calc_profit(r1_valid_df)

print(f"Region 1 Estimated Profit: ${r1_profit:.2f}")

Region 1 Estimated Profit: $24857120.52


In [62]:
r1_reserves = r1_profit / revenue_per_barrel

print(f"Region 1 Estimated Volume of Reserves According to Predictions: {int(r1_reserves)}")

Region 1 Estimated Volume of Reserves According to Predictions: 5523


In [63]:
r2_profit = calc_profit(r2_valid_df)

print(f"Region 2 Estimated Profit:${r2_profit:.2f}")

Region 2 Estimated Profit:$33217543.96


In [64]:
r2_reserves = r2_profit / revenue_per_barrel 

print(f"Region 2 Estimated Volume of Reserves According to Predicitons: {int(r2_reserves)}")

Region 2 Estimated Volume of Reserves According to Predicitons: 7381


#### Bootstrapping

In [65]:
r0_predicted = r0_valid_df['predicted']
r1_predicted = r1_valid_df['predicted']
r2_predicted = r2_valid_df['predicted']


r0_actual = r0_valid_df['actual_valid']
r1_acutal = r1_valid_df['acutal_valid']
r2_actual = r2_valid_df['actual_valid']

In [66]:
def bootstrap_sample(predicted, actual_valid, num_samples=1000, sample_size=200):
    profits = []
    
    for i in range(num_samples):
        indicies_sample = np.random.choice(len(predicted), size=sample_size, replace=True)
        predicted_sample = predicted.iloc[indicies_sample] if isinstance(predicted, pd.Series) else predicted[indices_sample]
        actual_sample = actual_valid.iloc[indicies_sample] if isinstance(actual_valid, pd.Series) else actual_valid[indices_sample]
        
        sorted_id = np.argsort(predicted_sample)[-200:]
        top_predictions = predicted_sample.iloc[sorted_id] if isinstance(predicted_sample, pd.Series) else predicted_sample[sorted_id]
        top_actual_values = actual_sample.iloc[sorted_id] if isinstance(actual_sample, pd.Series) else actual_sample[sorted_id]
        
        valid_df = pd.DataFrame({'predicted': top_predictions, 'actual': top_actual_values})
        
        profit = calc_profit(valid_df)
        
       
        profits.append(profit)
        
    return profits

In [67]:
bootstrap_0 = bootstrap_sample(r0_predicted, r0_actual)


r0_average = np.mean(bootstrap_0)

In [68]:
r0_average

-16665037.443385817

In [69]:
r0_confidence_interval = np.percentile(bootstrap_0, [2.5, 97.5])

r0_confidence_interval

array([-19549567.17359615, -13849104.45434633])

In [70]:
r0_risk_of_loss = np.mean((np.array(bootstrap_0) < 0)) * 100

r0_risk_of_loss

100.0

In [71]:
bootstrap_1 = bootstrap_sample(r1_predicted, r1_acutal)

In [72]:
r1_average = np.mean(bootstrap_1)

r1_average

-38091925.93095947

In [73]:
r1_risk_of_loss = np.mean((np.array(bootstrap_1) < 0)) * 100

r1_risk_of_loss

100.0

In [74]:
bootstrap_2 = bootstrap_sample(r2_predicted, r2_actual)

r2_average = np.mean(bootstrap_2)

r2_average

-14550400.99782231

In [75]:
r2_risk_of_loss = np.mean((np.array(bootstrap_2) < 0)) * 100

r2_risk_of_loss

100.0